include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,384 @@
+/*
+  SLB.Detail.Invoke
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#ifndef SLB_DETAIL_INVOKE_HPP
+#define SLB_DETAIL_INVOKE_HPP
+
+#include "config.hpp"
+#include "lib.hpp"
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace slb {
+namespace detail {
+namespace _invoke {
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T>
+struct is_reference_wrapper : std::false_type {};
+
+template <typename T>
+struct is_reference_wrapper<std::reference_wrapper<T>> : std::true_type {};
+
+////////////////////////////////////////////////////////////////////////////////
+struct ref_tag {};
+struct ref_wrapper_tag {};
+struct ptr_tag {};
+
+template <typename C, typename T1, typename Enable = void>
```

Comment:
Why `T1`? There seem to be no other `T`s here nor in the specializations.

Reply:
It refers to the type of the first (0th) argument in the `INVOKE` specification.

Reply:
Oki